lucille-core/src/main/java/com/kmwllc/lucille/connector/storageclient/BaseStorageClient.java

```diff
@@ -44,35 +45,75 @@ public abstract class BaseStorageClient implements StorageClient {
   private boolean initialized = false;
 
   public BaseStorageClient(Config cloudOptions) {
+    validateOptions(cloudOptions);
     this.cloudOptions = cloudOptions;
-    validateOptions();
-
     this.fileHandlers = new HashMap<>();
 
     // only matters for traversals
     this.maxNumOfPages = cloudOptions.hasPath(MAX_NUM_OF_PAGES) ? cloudOptions.getInt(MAX_NUM_OF_PAGES) : 100;
   }
 
+  /**
+   * Validate that the given cloudOptions are sufficient to construct an instance of this StorageClient. Throws an
+   * IllegalArgumentException if the cloudOptions do not contain the necessary information.
+   */
+  protected abstract void validateOptions(Config cloudOptions);
+
   @Override
   public final void init() throws IOException {
-    initializeStorageClient();
-    this.initialized = true;
+    if (!initialized) {
+      initializeStorageClient();
+      initialized = true;
+    }
   }
 
+  // Actually do the storage client specific initialization.
+  protected abstract void initializeStorageClient() throws IOException;
+
   @Override
   public final void shutdown() throws IOException {
-    this.initialized = false;
-    shutdownStorageClient();
+    if (initialized) {
+      initialized = false;
+      shutdownStorageClient();
+    }
   }
 
-  @Override
+  // Actually do the storage client specific shutdown.
+  protected abstract void shutdownStorageClient() throws IOException;
+
+  /**
+   * Returns whether this StorageClient is currently initialized successfully and not shutdown.
+   */
   public boolean isInitialized() {
     return this.initialized;
   }
 
-  protected abstract void initializeStorageClient() throws IOException;
+  @Override
+  public final void traverse(Publisher publisher, TraversalParams params) throws Exception {
+    if (!isInitialized()) {
+      throw new IOException("This StorageClient has not been initialized.");
```

Comment:
shouldn't be IOException because we haven't actually started an IO
I think IllegalStateException makes sense here